### PR TITLE
Feature - ThongNguyen - CustomView

### PR DESCRIPTION
**Pull request by mvn-thongnguyen-dn**
## Screenshot
iPhone 12 PROMAX
---

### CODE REVIEW

File: 04_View/ThongNT/View/View/CustomView/Avatar View/UserView.swift [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    func didTap(view: UserView, count: Int)
+}
+
+final class UserView: UIView {
+    
+    var count: Int = 0
+    var delegate: UserViewDelegate?
+    
+    @IBOutlet weak var userNameLabel: UILabel!
+    @IBOutlet weak var userImage: UIImageView!
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CustomView/Avatar View/UserView.swift:20]: Private

File: 04_View/ThongNT/View/View/CustomView/Avatar View/AvatarViewController.swift [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    private func setupUI() {
+        addUsers()
+    }
+    
+    private func addUser(username: String, userImage: String, x: CGFloat, y: CGFloat) {
+        guard let avatarView = Bundle.main.loadNibNamed("UserView", owner: self, options: nil)?.first as? UserView else { return }
+        avatarView.frame = CGRect(x: 150 + x, y: 150 + y, width: 100, height: 150)
+        avatarView.userNameLabel.text = username
+        avatarView.userImage.image = UIImage(named: "MonkeyImage")
+        avatarView.userImage.contentMode = .scaleToFill
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CustomView/Avatar View/AvatarViewController.swift:52]: Write a function in UserView to update data, please follow OOP

File: 04_View/ThongNT/View/View/CustomView/Avatar View/UserView.swift [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+import UIKit
+
+protocol UserViewDelegate {
+    func didTap(view: UserView, count: Int)
+}
+
+final class UserView: UIView {
+    
+    var count: Int = 0
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CustomView/Avatar View/UserView.swift:16]: Private

File: 04_View/ThongNT/View/View/CustomView/Avatar View/UserView.swift [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    var delegate: UserViewDelegate?
+    
+    @IBOutlet weak var userNameLabel: UILabel!
+    @IBOutlet weak var userImage: UIImageView!
+
+    
+
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CustomView/Avatar View/UserView.swift:23]: Need 1 line only to separate

File: 04_View/ThongNT/View/View/CustomView/Custom Slider/CustomSliderView.swift [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        super.layoutSubviews()
+        setupUI()
+    }
+    
+    //MARK: - Functions
+    func setThumbPosistionBaseOnSet() {
+        guard let dataSource = dataSource else { return }
+        let newPercent = dataSource.getChangeValue(view: self)
+        updateUI(with: newPercent)
+
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CustomView/Custom Slider/CustomSliderView.swift:41]: Remove unused line

File: 04_View/ThongNT/View/View/CustomView/Custom Slider/MySliderViewController.swift [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        
+        let origin = CGPoint(x: view.bounds.midX - 75, y: view.bounds.midY - 150)
+        sliderView.frame = CGRect(x: origin.x, y: origin.y, width: 150, height: 300)
+        sliderView.delegate = self
+        sliderView.dataSource = self
+        view.addSubview(sliderView)
+        self.sliderView = sliderView
+    }
+    
+    @IBAction func changeValue(_ sender: Any) {
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CustomView/Custom Slider/MySliderViewController.swift:39]: - Private
- Rename function, exactly what function do

File: 04_View/ThongNT/View/View/CustomView/Custom Slider/MySliderViewController.swift [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        if changeValue > 100 {
+            return 100
+        }
+        return changeValue
+    }
+    
+    func didChangeValue(view: CustomSliderView, value: Int) {
+        valueTextField.text = String(value)
+    }
+
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CustomView/Custom Slider/MySliderViewController.swift:62]: Remove unused line

File: 04_View/ThongNT/View/View/CustomView/Date Picker/DatePickerViewController.swift [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setupUI()
+    }
+    
+    //MARK: - Private functions
+    private func setupUI() {
+        loadNib()
+    }
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CustomView/Date Picker/DatePickerViewController.swift:26]: Why do you define a function to call another function?

File: 04_View/ThongNT/View/View/CustomView/Date Picker/DatePickerViewController.swift [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    
+    private func hideView() {
+        UIView.animate(withDuration: 1) {
+            self.datePickerView?.alpha = 0
+        }
+    }
+    
+    private func formatDate(from myDate: Date) {
+
+    }
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CustomView/Date Picker/DatePickerViewController.swift:56]: Remove empty function

File: 04_View/ThongNT/View/View/CustomView/Date Picker/DatePickerViewController.swift [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+// MARK: - Extension Delegate
+extension DatePickerViewController: DataPickerViewDelegate {
+    func sendDate(in view: DatePickerCustomView, mydate: Date, isHide: Bool) {
+        showDateTextField.text = "\(mydate)"
+        if isHide {
+            hideView()
+        } else {
+            showView()
+        }
```
Comment (mvn-bienle-dn) [inline 04_View/ThongNT/View/View/CustomView/Date Picker/DatePickerViewController.swift:71]: Can refactor with:
isHide ? hideView() : showView()

**Review by mvn-bienle-dn — COMMENTED** [review]
@mvn-thongnguyen-dn Please check my comments.
Thank you!

File: 04_View/ThongNT/View/View/CustomView/Date Picker/DatePickerViewController.swift (hunk 8 above)
Comment (mvn-thongnguyen-dn) [reply 04_View/ThongNT/View/View/CustomView/Date Picker/DatePickerViewController.swift:26]: I wrote that to practice with encapsulation. In future, if i want to add more config, just write it's in this function.

**Comment by mvn-thongnguyen-dn** [comment]
@mvn-bienle-dn Please check my fixes. 
Thanks you.

**Review by mvn-bienle-dn — APPROVED** [review]
LGTM!

**Pull request merged**